lib/charms/grafana_agent/v0/cos_machine.py

```diff
@@ -0,0 +1,220 @@
+        self._logs_slots = logs_slots or []
+        self._dashboard_dirs = dashboard_dirs
+        self._refresh_events = refresh_events or [self._charm.on.config_changed]
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_joined, self.update_relation_data)
+        self.framework.observe(events.relation_changed, self.update_relation_data)
+        for event in self._refresh_events:
+            self.framework.observe(event, self.update_relation_data)
+
+    def update_relation_data(self, event):
+        if isinstance(event, RelationEvent):
+            relations = [event.relation]
+        else:
+            relations = self._charm.model.relations[self._relation_name]
+
+        for relation in relations:
+            relation.data = self._generate_updated_data()
+
+    def _generate_updated_data(self) -> str:
+        """Collate the data for each nested databag and return it."""
+        data = {
+            "metrics": {
+                "scrape_jobs": self._scrape_jobs,
+                "alert_rules": self._metrics_alert_rules,
+            },
+            "logs": {
+                "targets": self._logs_slots,
+                "alert_rules": self._log_alert_rules,
+            },
+            # Probably doesn't actually need to be nested
+            "dashboards": {
+                "dashboards": self._dashboards,
+            },
+        }
+
+        return json.dumps(data)
+
+    def _scrape_jobs(self) -> List[Dict]:
+        """Return a prometheus_scrape-like data structure for jobs."""
+        job_name_prefix = self._charm.app.name
+        return [
+            {"job_name": f"{job_name_prefix}_{key}", **endpoint}
+            for key, endpoint in enumerate(self._metrics_endpoints)
+        ]
+
+    def _metrics_alert_rules(self) -> Dict:
+        """Use (for now) the prometheus_scrape AlertRules to initialize this."""
+        alert_rules = MetricsAlerts()
+        alert_rules.add_path(self.metrics_rules_dir, recursive=self._recursive)
+        return alert_rules.as_dict()
+
+    def _log_alert_rules(self) -> Dict:
+        """Use (for now) the loki_push_api AlertRules to initialize this."""
+        alert_rules = LogAlerts()
+        alert_rules.add_path(self.log_rules_dir, recursive=self._recursive)
+        return alert_rules.as_dict()
+
+    def _dashboards(self) -> List[str]:
+        dashboards = []
+        for d in self._dashboard_dirs:
+            for path in Path(d).glob("*"):
+                dashboards.append(self._encode_dashboard_content(path.read_bytes()))
+
+        return dashboards
+
+    def _encode_dashboard_content(self, content: Union[str, bytes]) -> str:
+        if isinstance(content, str):
+            content = bytes(content, "utf-8")
+
+        return base64.b64encode(lzma.compress(content)).decode("utf-8")
+
+
+class CosMachineDataChanged(EventBase):
+    """Event emitted when a `CosMachineProvider` joins or updates data."""
```

Comment:
Is this ever going to do anything more than re-emit after relation-changed? If not, I think we should just not have it. I know we do this in our other libraries but I think it is kind of useless.

Reply:
It feels more meaningful to observe lib-emitted events than to observe relation events, no?
It's part of the library API. And some day we might extend the API by attaching attributes to this event, etc...
Feels more extensible than not having it.

Reply:
It just sort of feels like we are making the library more obfuscated. We do do this with all of our libraries and as far as I am aware, we have never actually used it for anything.

Reply:
It's an established pattern. I think we should continue doing it.

Reply:
mm, worth thinking about it for sure